### PR TITLE
Additional work to release unused PreviewWorkspace memory

### PR DESCRIPTION
**Pull request by sharwell**
Local heap dump shows 6.8GB memory retained by unused PreviewWorkspace instances. The GC root is linked through event handlers on GlobalOptionService. Since this has been a problem in the past, I took a multi-approach solution:

1. More aggressively track and dispose PreviewWorkspace instances. They are stored in `ReferenceCountedDisposable<>` from the point of initial creation, with ownership transferred by a pattern of duplication of a reference followed by releasing the original. Outside of exceptions thrown by the constructor(s), instances should now be disposed in the event of any exception.
2. Force the use of a weak event handler pattern for GlobalOptionService. In the event a PreviewWorkspace fails to be released, it will no longer be GC rooted through this event.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs [hunk 1]
```diff
@@ -46,9 +47,25 @@ private class LazyToolTip : ForegroundThreadAffinitizedObject
...
+                    _disposableToolTip?.Dispose();
+                    _disposableToolTip = null;
+                }
+                catch (Exception ex) when (FatalError.ReportAndCatch(ex))
+                {
+                    // Avoid propagating the exception within a UI cleanup layer
+                }
+            }
+
             private void OnToolTipOpening(object sender, ToolTipEventArgs e)
```
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs:69]: 📝 If an exception was no the original source of a PreviewWorkspace leaking, this is the second most likely case. This type assumes that there is a strict matching of calls to `OnToolTipOpening` and `OnToolTipClosing`. If there is ever a case where `OnToolTipOpening` is called more times than `OnToolTipClosing`, the additional calls would have leaked a `PreviewWorkspace` instance.

File: src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs [hunk 2]
```diff
@@ -68,13 +69,13 @@ public DisposableToolTip CreateDisposableToolTip(Document document, ITextBuffer
             // 
             // our underlying preview tagger and mechanism to attach tagger to associated buffer of
             // opened document will light up automatically
-            var workspace = new PreviewWorkspace(document.Project.Solution);
-            workspace.OpenDocument(document.Id, textBuffer.AsTextContainer());
+            using var workspace = PreviewWorkspace.Create(document.Project.Solution);
+            workspace.Target.OpenDocument(document.Id, textBuffer.AsTextContainer());
 
-            return new DisposableToolTip(toolTip, workspace);
+            return DisposableToolTip.Create(toolTip, workspace);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs:75]: visually, this just looks weird.  It looks like you're returning something that is holding onto something that is explicitly being disposed.  needs comments to explain why this is ok.

File: src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs [hunk 3]
```diff
@@ -18,17 +19,17 @@ internal partial class ContentControlService
         /// <summary>
         /// Class which allows us to provide a delay-created tooltip for our reference entries.
         /// </summary>
-        private class LazyToolTip : ForegroundThreadAffinitizedObject
+        private class LazyToolTip : ForegroundThreadAffinitizedObject, IDisposable
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs:22]: perhaps LazyDisposableToolTip?

File: src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs [hunk 4]
```diff
@@ -58,8 +75,11 @@ private void OnToolTipOpening(object sender, ToolTipEventArgs e)
                     Debug.Assert(_element.ToolTip == this);
                     Debug.Assert(_disposableToolTip == null);
 
+                    // We don't expect _disposableToolTip to be non-null here, but we still make sure it's not leaking
+                    _disposableToolTip?.Dispose();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs:79]: what about just not doing work in that case?  if hte tooltip was already created, we could just immediately exit?

File: src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs [hunk 5]
```diff
@@ -287,42 +287,42 @@ private async ValueTask<DifferenceViewerPreview> CreateAddedDocumentPreviewViewC
...
             // Create PreviewWorkspace around the buffer to be displayed in the diff preview
             // so that all IDE services (colorizer, squiggles etc.) light up in this buffer.
-            var rightWorkspace = new PreviewWorkspace(document.Project.Solution);
-            rightWorkspace.OpenDocument(document.Id, newBuffer.AsTextContainer());
+            using var rightWorkspace = PreviewWorkspace.Create(document.Project.Solution);
+            rightWorkspace.Target.OpenDocument(document.Id, newBuffer.AsTextContainer());
 
 #pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task (containing method uses JTF)
             return await CreateAddedDocumentPreviewViewCoreAsync(newBuffer, rightWorkspace, document, zoomLevel, cancellationToken);
 #pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs:295]: same points about returning something that captures a local which appears to be disposed when we get to end of scope.

File: src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs [hunk 6]
```diff
@@ -670,21 +672,15 @@ private async ValueTask<ITextBuffer> CreateNewPlainTextBufferAsync(TextDocument
...
             if (_editorOptionsService.GlobalOptions.GetOption(SolutionCrawlerRegistrationService.EnableSolutionCrawler))
             {
-                leftWorkspace?.EnableSolutionCrawler();
-                rightWorkspace?.EnableSolutionCrawler();
+                leftWorkspace?.Target.EnableSolutionCrawler();
+                rightWorkspace?.Target.EnableSolutionCrawler();
             }
 
-            return new DifferenceViewerPreview(diffViewer);
+            return differenceViewer.TryAddReference()!;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs:683]: for my own edification, this should always succeed, yes? (i presume that's also why we have the `!`).  I feel ike it would be nice on these RefCountedDisposables to have a method (or extension) that adds and asserts/throws if that somehow failed (Similar to GetRequiredDocument vs GetDocument).

File: src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs [hunk 7]
```diff
@@ -5,24 +5,39 @@
...
-        public DifferenceViewerPreview(IDifferenceViewer viewer)
+        private DifferenceViewerPreview(
+            IDifferenceViewer viewer,
+            ReferenceCountedDisposable<PreviewWorkspace> leftWorkspace,
+            ReferenceCountedDisposable<PreviewWorkspace> rightWorkspace)
         {
             Contract.ThrowIfNull(viewer);
             _viewer = viewer;
+            LeftWorkspace = leftWorkspace?.TryAddReference();
+            RightWorkspace = rightWorkspace?.TryAddReference();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs:27]: why `?.`?  You never are passed null instances in right?

Also, is this another location where we could use the asserting AddRef methods?

File: src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs [hunk 8]
```diff
@@ -33,6 +48,9 @@ public void Dispose()
             }
 
             _viewer = null;
+
+            LeftWorkspace?.Dispose();
+            RightWorkspace?.Dispose();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs:53]: no need for `?.` afaict.

File: src/EditorFeatures/Test/Options/GlobalOptionsTests.cs [hunk 9]
```diff
@@ -99,6 +95,12 @@ public void SetGlobalOption(OptionKey2 optionKey, object? value)
         public bool SetGlobalOptions(ImmutableArray<KeyValuePair<OptionKey2, object?>> options)
             => throw new NotImplementedException();
 
+        public void AddOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
+            => throw new NotImplementedException();
+
+        public void RemoveOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
+            => throw new NotImplementedException();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test/Options/GlobalOptionsTests.cs:102]: would it be possible to make two PRs?  one for the refcounted stuff, and one for the weak event handlerS?

File: src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs [hunk 10]
```diff
@@ -185,18 +186,29 @@ protected override IList<System.Windows.Documents.Inline> CreateLineTextInlines(
...
                     // this lazy tooltip causes whole solution to be kept in memory until find all reference result gets cleared.
                     // solution is never supposed to be kept alive for long time, meaning there is bunch of conditional weaktable or weak reference
                     // keyed by solution/project/document or corresponding states. this will cause all those to be kept alive in memory as well.
                     // probably we need to dig in to see how expensvie it is to support this
                     var controlService = _excerptResult.Document.Project.Solution.Services.GetRequiredService<IContentControlService>();
-                    controlService.AttachToolTipToControl(content, () =>
+                    var disposable = controlService.AttachToolTipToControl(content, () =>
                         CreateDisposableToolTip(_excerptResult.Document, _excerptResult.Span));
 
+                    disposeCallback += () =>
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs:201]: wow.  that this is being `+=` is subtle.  Can you doc htat?  I was def confused a bit. :)

File: src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs [hunk 11]
```diff
@@ -185,18 +186,29 @@ protected override IList<System.Windows.Documents.Inline> CreateLineTextInlines(
...
+                    {
+                        try
+                        {
+                            disposable?.Dispose();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs:205]: why `?.` here?

File: src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs [hunk 12]
```diff
@@ -31,6 +36,23 @@ public TableEntriesSnapshot(ImmutableList<Entry> entries, int versionNumber)
 
             public override int Count => _entries.Count;
 
+            public override void Dispose()
+            {
+                foreach (var callback in _disposeCallbacks)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs:41]: nit: consider threading assert.

File: src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs [hunk 13]
```diff
@@ -20,6 +24,7 @@ private class TableEntriesSnapshot : WpfTableEntriesSnapshotBase
             private readonly int _versionNumber;
 
             private readonly ImmutableList<Entry> _entries;
+            private readonly List<Action> _disposeCallbacks = new();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs:27]: docs plz.

File: src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs:201]: dumb question.  i see what you're doing here, creating a side-table of disposables that go away when the snapshot goes away.  Is it not possible though to just reutrn FrameworkElements that do this cleanup once they themselves are disposed?

File: src/VisualStudio/Core/Def/Preview/PreviewEngine.cs [hunk 14]
```diff
@@ -75,9 +70,25 @@ public PreviewEngine(IThreadingContext threadingContext, string title, string he
...
+            IVsImageService2 imageService,
+            bool showCheckBoxes = true)
+        {
+            return new ReferenceCountedDisposable<PreviewEngine>(new PreviewEngine(threadingContext, title, helpString, description, topLevelItemName, topLevelGlyph, newSolution, oldSolution, componentModel, imageService, showCheckBoxes));
+        }
+
+        public void Dispose()
         {
-            _updater?.CloseWorkspace();
+            _updater?.Dispose();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Preview/PreviewEngine.cs:91]: why `?.`?

File: src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs [hunk 15]
```diff
@@ -15,22 +16,25 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.Preview
 {
-    internal partial class PreviewUpdater : ForegroundThreadAffinitizedObject
+    internal partial class PreviewUpdater : ForegroundThreadAffinitizedObject, IDisposable
     {
-        private PreviewDialogWorkspace? _previewWorkspace;
+        private ReferenceCountedDisposable<PreviewDialogWorkspace>? _previewWorkspace;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs:21]: wow... this is lazily created and assigned.... i hate this :D (same problem before your code).

File: src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs [hunk 16]
```diff
@@ -66,17 +72,27 @@ private void UpdateBuffer(TextDocument document)
...
-                _previewWorkspace.CloseDocument(currentDocument, currentDocumentText);
+                previewWorkspace.Target.CloseDocument(currentDocument, currentDocumentText);
+            }
+
+            if (previewWorkspace is null)
+            {
+                var newWorkspace = PreviewDialogWorkspace.Create(document.Project.Solution);
+                if (Interlocked.CompareExchange(ref _previewWorkspace, newWorkspace, null) is not null)
+                    newWorkspace.Dispose();
             }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs:86]: i'm def confused about why this is all written this way.  i would presume the PreviewUpdater is for something like the options page.  and deferring the whole workspace stuff just seems totally unnecessary (And makes things so contorted and hard to follow).

File: src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs [hunk 17]
```diff
@@ -66,17 +72,27 @@ private void UpdateBuffer(TextDocument document)
...
 
-            _previewWorkspace ??= new PreviewDialogWorkspace(document.Project.Solution);
+            RoslynDebug.AssertNotNull(_previewWorkspace);
+
+            using var previewWorkspace2 = _previewWorkspace.TryAddReference();
+            RoslynDebug.AssertNotNull(previewWorkspace2);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs:91]: RequireAddReference :)

File: src/Workspaces/Core/Portable/Options/GlobalOptionService.cs [hunk 18]
```diff
@@ -236,27 +243,103 @@ public bool RefreshOption(OptionKey2 optionKey, object? newValue)
...
+
+            ImmutableInterlocked.Update(
+                ref _weakHandlers,
+                static (weakHandlers, handler) =>
+                {
+                    return weakHandlers
+                        .RemoveAll(WeakReferenceExtensions.IsNull)
+                        .Add(new WeakReference<EventHandler<OptionChangedEventArgs>>(handler));
+                },
+                handler);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/GlobalOptionService.cs:270]: seems odd to need ImmutableInterlocked *and* a `_gate`.  Can we just do all things atomically under the one case?

File: src/Workspaces/Core/Portable/Options/IGlobalOptionService.cs [hunk 19]
```diff
@@ -43,21 +43,19 @@ internal interface IGlobalOptionService : IOptionsReader
...
         /// Atomically sets the values of specified global options. The option values are persisted.
-        /// Triggers <see cref="OptionChanged"/>.
+        /// Triggers option changed event for handlers registered with <see cref="AddOptionChangedHandler"/>.
         /// </summary>
         /// <remarks>
         /// Returns true if any option changed its value stored in the global options.
         /// </remarks>
         bool SetGlobalOptions(ImmutableArray<KeyValuePair<OptionKey2, object?>> options);
 
-        event EventHandler<OptionChangedEventArgs>? OptionChanged;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/IGlobalOptionService.cs:59]: unclear if this will impact partners. need to probably check on TS at least since they aren't EA on everything.

File: src/Workspaces/Core/Portable/Options/GlobalOptionService.cs (hunk 18 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Options/GlobalOptionService.cs:270]: The common case is reading from this list. The code is structured so the uncommon case is the only one that needs a lock.

File: src/VisualStudio/Core/Def/Preview/PreviewEngine.cs [hunk 20]
```diff
@@ -218,15 +229,27 @@ public void UpdatePreview(DocumentId documentId, SpanChange spanSource)
             var document = updatedSolution.GetTextDocument(documentId);
             if (document != null)
             {
-                _updater.UpdateView(document, spanSource);
+                using var updater = _updater?.TryAddReference();
+                RoslynDebug.AssertNotNull(updater);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewEngine.cs:233]: 📝 It would just be `AddReference()`

File: src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs:75]: This is _the_ correct pattern for correctly transferring ownership of a `ReferenceCountedDisposable<T>`. An RCD passed by value is only assured to be alive until the end of the call. The callee must obtain its own reference if it needs to store the value to a field. The behavior of the caller depends on the source of the RCD: if the caller created the RCD (as is the case here) it must release it before returning. If the caller received the RCD via an argument, it doesn't own that instance and can just return.

File: src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs (hunk 3 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs:22]: ➡️ Keeping the current name because 1) it's impossible to create an instance of this without realizing it's disposable (the `Create` method returns RCD) and 2) there is no corresponding non-disposable type.

File: src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs:75]: > This is the correct pattern for correctly transferring ownership of a ReferenceCountedDisposable<T>

literally nothing in the code states taht it's an RCD though.  It just looks like a bog standard captured disposable.  RCD is fine here because the using is only affecting the ref-count. And, when that is clear, is totally ok.  But in cases like this, there's no indication of ref-counting at all, so it immediatley jumps out as looking problematic.

File: src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs (hunk 4 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/QuickInfo/LazyToolTip.cs:79]: ⏱️ Sure that seems reasonable

File: src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs (hunk 5 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs:295]: ➡️ Covered by https://github.com/dotnet/roslyn/pull/68054/files#r1182949296

File: src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs:683]: ⏱️ Will add a `AddReference()` method

File: src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs (hunk 7 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs:27]: Multiple locations pass in null

Comment (sharwell) [reply src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs]: ➡️ These can be null. See 6a08a737958648f643d2aed618603b02c895e1c3

File: src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs (hunk 8 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs:51]: ➡️ These can be null. See 6a08a737958648f643d2aed618603b02c895e1c3

File: src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Preview/AbstractPreviewFactoryService.cs]: ➡️ Implemented in b228912c0d883c2f39f5413ace7f46cbd9c279b4

File: src/EditorFeatures/Test/Options/GlobalOptionsTests.cs (hunk 9 above)
Comment (sharwell) [reply src/EditorFeatures/Test/Options/GlobalOptionsTests.cs]: ➡️ Extracted #68062

File: src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs (hunk 10 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs:201]: 📝 FrameworkElement is not disposable, and WPF does not notify on final release

File: src/VisualStudio/Core/Def/Preview/PreviewEngine.cs (hunk 14 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewEngine.cs:91]: ➡️ _updater is lazily created

File: src/VisualStudio/Core/Def/Preview/PreviewEngine.cs [hunk 21]
```diff
@@ -218,15 +229,27 @@ public void UpdatePreview(DocumentId documentId, SpanChange spanSource)
             var document = updatedSolution.GetTextDocument(documentId);
             if (document != null)
             {
-                _updater.UpdateView(document, spanSource);
+                using var updater = _updater?.TryAddReference();
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewEngine.cs]: ⏱️ Switching to Contract.ThrowIfNull at this location

File: src/VisualStudio/Core/Def/Preview/PreviewEngine.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewEngine.cs:233]: ⏱️ Switching to AddReference at this location

File: src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs (hunk 16 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs:86]: ➡️ Planning to leave this one as-is since it's the same as before

File: src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs (hunk 17 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs]: ➡️ Part of b228912c0d883c2f39f5413ace7f46cbd9c279b4

File: src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs (hunk 15 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Preview/PreviewUpdater.cs:21]: ➡️ Planning to leave this one as-is since it's the same as before

File: src/Workspaces/Core/Portable/Options/GlobalOptionService.cs (hunk 18 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Options/GlobalOptionService.cs]: ➡️ Moved to #68062

File: src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs (hunk 12 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs:41]: ⏱️  Will remove affinity further down stream

File: src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs (hunk 8 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Preview/DifferenceViewerPreview.cs:51]: ➡️ Covered by 6a08a737958648f643d2aed618603b02c895e1c3

File: src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs (hunk 13 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/FindReferences/TableEntriesSnapshot.cs:31]: ➡️  Small amount of docs added

File: src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs (hunk 10 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/FindReferences/Entries/DocumentSpanEntry.cs:201]: ➡️ Documented the use of the `+=` operator

File: src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/Core.Wpf/QuickInfo/ContentControlService.cs:75]: ➡️ Renamed `Create` methods that return RCD to `CreateReferenceCounted`
